pkg/manifest/objects_test.go

```diff
@@ -0,0 +1,249 @@
+				"command": [
+					"sh",
+					"-c",
+					"echo Hello Kubernetes! && sleep 3600"
+				]
+			}
+		]
+	}
+}`
+	testService := `{
+	"apiVersion": "v1",
+	"kind": "Service",
+	"metadata": {
+		"name": "my-service",
+		"namespace": "test-apps"
+	},
+	"spec": {
+		"selector": {
+			"app": "MyApp"
+		},
+		"ports": [
+			{
+				"protocol": "TCP",
+				"port": 80,
+				"targetPort": 9376
+			}
+		]
+	}
+}`
+
+	parseJSONToK8sObjectTests := []struct {
+		objString     string
+		wantGroup     string
+		wantKind      string
+		wantName      string
+		wantNamespace string
+	}{
+		{testDeployment, "apps", "Deployment", "nginx-deployment", "test-apps"},
+		{testPod, "", "Pod", "myapp-pod", "test-apps"},
+		{testService, "", "Service", "my-service", "test-apps"},
+	}
+
+	for _, tt := range parseJSONToK8sObjectTests {
+		k8sObj, err := ParseJSONToK8sObject([]byte(tt.objString))
+		if err != nil {
+			k8sObjStr, err := k8sObj.YAMLDebugString()
+			if err != nil {
+				if k8sObj.Group != tt.wantGroup {
+					t.Errorf("got group %s for k8s object %s, want %s", k8sObj.Group, k8sObjStr, tt.wantGroup)
+				}
+				if k8sObj.Group != tt.wantGroup {
+					t.Errorf("got kind %s for k8s object %s, want %s", k8sObj.Kind, k8sObjStr, tt.wantKind)
+				}
+				if k8sObj.Name != tt.wantName {
+					t.Errorf("got name %s for k8s object %s, want %s", k8sObj.Name, k8sObjStr, tt.wantName)
+				}
+				if k8sObj.Namespace != tt.wantNamespace {
+					t.Errorf("got group %s for k8s object %s, want %s", k8sObj.Namespace, k8sObjStr, tt.wantNamespace)
+				}
+			}
+		}
+	}
+}
+
+func TestParseYAMLToK8sObject(t *testing.T) {
+	testManifest := `apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  namespace: test-apps
+  labels:
+    app: nginx
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: nginx:1.7.9
+        ports:
+        - containerPort: 80
+---
+kind: Pod
+metadata:
+  name: myapp-pod
+  namespace: test-apps
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp-container
+    image: busybox
+    command: ['sh', '-c', 'echo Hello Kubernetes! && sleep 3600']
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+  namespace: test-apps
+spec:
+  selector:
+    app: MyApp
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 9376`
+
+	parseK8sObjectsFromYAMLManifestTests := []struct {
+		objString string
+		hashes    []string
+	}{
+		{testManifest, []string{"Deployment:test-apps:nginx-deployment", "Pod:test-apps:myapp-pod", "Service:test-apps:my-service"}},
+	}
+
+	for _, tt := range parseK8sObjectsFromYAMLManifestTests {
+		k8sObjs, err := ParseK8sObjectsFromYAMLManifest(tt.objString)
+		if err != nil {
+			for i, k8sObj := range k8sObjs {
+				gotHash := k8sObj.Hash()
+				k8sObjStr, err := k8sObj.YAMLDebugString()
+				if err != nil {
+					if k8sObj.Hash() != tt.hashes[i] {
```

Comment:
This is only testing the hashes - it would be great to test the round trip for the objects e.g. start with a a map of hash:object YAML string, join these together into a single string, then parse back and check and hash:object yamls are the same. You can use something like this: https://github.com/istio/operator/blob/d98093ec44c1d59a6c42865b2dd50e988e42f758/pkg/patch/patch_test.go#L433
(actually you can just check that output from YAMLDiff is "" instead of IsYAMLEqual)

Reply:
Good idea, done.